### ISSUE TITLE
Three tests fails on OpenIndiana

### ISSUE THREAD

**Issue by mtelka** [zaneb/autopage#4]
The following three tests fails when I test autopage 0.5.1 on OpenIndiana:
```
=================================== FAILURES ===================================
_____________ InvokePagerTest.test_interrupt_at_end_after_complete _____________

self = <autopage.tests.test_end_to_end.InvokePagerTest testMethod=test_interrupt_at_end_after_complete>

    def test_interrupt_at_end_after_complete(self) -> None:
        num_lines = 100
        with isolation.isolate(finite(num_lines)) as env:
            pager = isolation.PagerControl(env)

            while pager.advance():
                continue

            self.assertEqual(num_lines, pager.total_lines())

            for i in range(100):
                env.interrupt()

>           self.assertEqual(0, pager.quit())

autopage/tests/test_end_to_end.py:178:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.9/vendor-packages/teamcity/diff_tools.py:33: in _patched_equals
    old(self, first, second, msg)
E   AssertionError: 0 != 24
___________________ LineBufferFromInputTest.test_default_tty ___________________

self = <autopage.tests.test_line_buffer_from_input.LineBufferFromInputTest testMethod=test_default_tty>

    def test_default_tty(self) -> None:
        with sinks.TTYFixture() as inp:
            with fixtures.MonkeyPatch('sys.stdin', inp.stream):
>               self.assertTrue(autopage.line_buffer_from_input())
E               AssertionError: False is not true

autopage/tests/test_line_buffer_from_input.py:34: AssertionError
_______________________ LineBufferFromInputTest.test_tty _______________________

self = <autopage.tests.test_line_buffer_from_input.LineBufferFromInputTest testMethod=test_tty>

    def test_tty(self) -> None:
        with sinks.TTYFixture() as inp:
>           self.assertTrue(autopage.line_buffer_from_input(inp.stream))
E           AssertionError: False is not true

autopage/tests/test_line_buffer_from_input.py:25: AssertionError
=========================== short test summary info ============================
FAILED autopage/tests/test_end_to_end.py::InvokePagerTest::test_interrupt_at_end_after_complete
FAILED autopage/tests/test_line_buffer_from_input.py::LineBufferFromInputTest::test_default_tty
FAILED autopage/tests/test_line_buffer_from_input.py::LineBufferFromInputTest::test_tty
======================== 3 failed, 151 passed in 45.26s ========================
```

**Comment by mtelka** [comment]
In addition to above three tests the following test fails in about half of test runs:
```
___________ InvokePagerTest.test_interrupt_in_middle_after_complete ____________

self = <autopage.tests.test_end_to_end.InvokePagerTest testMethod=test_interrupt_in_middle_after_complete>

    def test_interrupt_in_middle_after_complete(self) -> None:
        num_lines = 100
        with isolation.isolate(finite(num_lines)) as env:
            pager = isolation.PagerControl(env)
    
            self.assertEqual(MAX_LINES_PER_PAGE, pager.advance())
    
            for i in range(100):
                env.interrupt()
    
>           self.assertEqual(MAX_LINES_PER_PAGE, pager.quit())

autopage/tests/test_end_to_end.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.9/vendor-packages/teamcity/diff_tools.py:33: in _patched_equals
    old(self, first, second, msg)
E   AssertionError: 23 != 24
```

**Comment by zaneb** [comment]
The failure of the 2 test_line_buffer_from_input tests seems inexplicable, since it can only [mean](https://github.com/zaneb/autopage/blob/v0.5.1/autopage/__init__.py#L266) that `seekable()` is returning True for a stream [opened](https://github.com/zaneb/autopage/blob/v0.5.1/autopage/tests/sinks.py#L25-L26) from `openpty()`.

I have to think this is something to do with OpenIndiana itself - it's even possible there is an issue with CPython's implementation os `seekable()` on Solaris/Illumos.

**Comment by zaneb** [comment]
Some of the end-to-end tests unfortunately suffer from flakes due to timing that are very difficult to track down. There could also be differences in the behaviour of `less` on different OSes that would result in the tests behaving differently even if the actual library is not.

**Comment by zaneb** [comment]
Solaris man pages (e.g. for [`fflush()`](https://docs.oracle.com/cd/E86824_01/html/E54766/fflush-3c.html)) claim that:

> A stream is seekable if the underlying file is not a pipe, FIFO, socket, or TTY device.

(which would be consistent with Linux/glibc), but running a simple C program (on OpenIndiana Hipster 2023.05) shows that no error is returned from `fseek()` on a TTY:

```
#include <stdio.h>
#include <errno.h>

int main(int argc, char **argv) {
    int r = fseek(stdin, 10, 0);
    printf("result=%d\nerr=%d\n", r, errno);
    return 0;
}
```

Naturally Python does not presume to know better than the OS, so `seekable()` returns True even on sys.stdin itself when called from a terminal (as well as the devices we have created with `openpty()` in the unit tests). However, since only TTYs are affected, we can detect and correct for this.

The other issues not related to TTYs seekability appear to have been resolved. The current version of less in 2023.05 is v643 - the failure in `test_interrupt_at_end_after_complete` noted above, as well as the one not mentioned here (but now reproducible with 0.5.1) in #5, are fixed by de8bd4279630ea4b09e86d1d3ac443e0c7e29f6b. The flaky tests are much improved by 69786a4190f94fd4327262ae046747954ed7f0df.

**Comment by mtelka** [comment]
Thank you for the analysis.  I reported the issue to illumos: https://www.illumos.org/issues/15994
